core/src/main/java/org/elasticsearch/cluster/routing/allocation/command/AllocatePrimaryAllocationCommand.java

```diff
@@ -0,0 +1,297 @@
+import org.elasticsearch.cluster.routing.RoutingNodes;
+import org.elasticsearch.cluster.routing.ShardRouting;
+import org.elasticsearch.cluster.routing.allocation.RerouteExplanation;
+import org.elasticsearch.cluster.routing.allocation.RoutingAllocation;
+import org.elasticsearch.cluster.routing.allocation.allocator.ShardsAllocators;
+import org.elasticsearch.cluster.routing.allocation.decider.Decision;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.xcontent.ToXContent;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.common.xcontent.XContentParser;
+import org.elasticsearch.gateway.AsyncShardFetch;
+import org.elasticsearch.gateway.TransportNodesListGatewayStartedShards;
+import org.elasticsearch.index.shard.ShardId;
+
+import java.io.IOException;
+import java.util.Map;
+
+/**
+ * Allocates an unassigned primary shard. The node to assign to can be either directly chosen
+ * or by providing an allocation id which then searches for a node that has a shard copy available
+ * with that allocation id. In contrast to {@link AllocateAllocationCommand}, this command only allocates
+ * if a shard copy is available on the selected node. However, it can still lead to data loss if the shard
+ * copy is out of date. The <tt>ignoreDeciders</tt> flag can be used to allocate the shard to a node even if the
+ * allocation deciders say no; use this with extreme care!
+ */
+public class AllocatePrimaryAllocationCommand implements AllocationCommand {
+
+    public static final String NAME = "allocate_primary";
+    public static final String RECOVER_PRIMARY_ALLOCATION_COMMAND = "allocate_primary_allocation_command";
+
+    public static class Factory implements AllocationCommand.Factory<AllocatePrimaryAllocationCommand> {
+
+        @Override
+        public AllocatePrimaryAllocationCommand readFrom(StreamInput in) throws IOException {
+            return new AllocatePrimaryAllocationCommand(ShardId.readShardId(in), in.readOptionalString(), in.readOptionalString(), in.readBoolean());
+        }
+
+        @Override
+        public void writeTo(AllocatePrimaryAllocationCommand command, StreamOutput out) throws IOException {
+            command.shardId.writeTo(out);
+            out.writeOptionalString(command.node);
+            out.writeOptionalString(command.allocationId);
+            out.writeBoolean(command.ignoreDeciders);
+        }
+
+        @Override
+        public AllocatePrimaryAllocationCommand fromXContent(XContentParser parser) throws IOException {
+            String index = null;
+            int shardId = -1;
+            String nodeId = null;
+            String allocationId = null;
+            boolean ignoreDeciders = false;
+
+            String currentFieldName = null;
+            XContentParser.Token token;
+            while ((token = parser.nextToken()) != XContentParser.Token.END_OBJECT) {
```

Comment:
I essentially copy pasted this stuff from the `AllocateAllocationCommand`. You are right though, I should transform it to use `ObjectParser`.